test/Microsoft.ML.Core.Tests/UnitTests/TestHosts.cs

```diff
@@ -55,23 +55,57 @@ public void TestCancellation()
                     do
                     {
                         index = rand.Next(hosts.Count);
-                    } while (hosts.ElementAt(index).Item1.IsCancelled || hosts.ElementAt(index).Item2 < 3);
-                    hosts.ElementAt(index).Item1.StopExecution();
+                    } while ((hosts.ElementAt(index).Item1 as ICancelableHost).IsCanceled || hosts.ElementAt(index).Item2 < 3);
+                    (hosts.ElementAt(index).Item1 as ICancelableHost).CancelExecution();
                     rootHost = hosts.ElementAt(index).Item1;
                     queue.Enqueue(rootHost);
                 }
                 addThread.Join();
                 while (queue.Count > 0)
                 {
                     var currentHost = queue.Dequeue();
-                    Assert.True(currentHost.IsCancelled);
+                    Assert.True((currentHost as ICancelableHost).IsCanceled);
 
                     if (children.ContainsKey(currentHost))
                         children[currentHost].ForEach(x => queue.Enqueue(x));
                 }
             }
         }
 
+        [Fact]
+        public void TestCancellationApi()
+        {
+            IHostEnvironment env = new MLContext(seed: 42);
+            var mainHost = env.Register("Main");
+            var children = new ConcurrentDictionary<IHost, List<IHost>>();
+            var hosts = new BlockingCollection<Tuple<IHost, int>>();
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("1"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("2"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("3"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("4"), 1));
+            hosts.Add(new Tuple<IHost, int>(mainHost.Register("5"), 1));
+            var addThread = new Thread(
```

Comment:
Why are we using a different thread here? I don't see any value in it. You are starting a thread, and then immediately joining it. You might as well just run the `for` loop below on the same thread.